src/Stratis.Bitcoin.Features.BlockStore/BlockStoreQueue.cs

```diff
@@ -179,56 +179,67 @@ public async Task InitializeAsync()
         }
 
         /// <inheritdoc/>
-        public Task<Transaction> GetTransactionByIdAsync(uint256 trxid)
+        public async Task<Transaction> GetTransactionByIdAsync(uint256 trxid)
         {
             // Only look for transactions if they're indexed.
             if (!this.storeSettings.TxIndex)
-                return Task.FromResult(default(Transaction));
+                return default(Transaction);
 
-            lock (this.blocksCacheLock)
-            {
-                foreach (ChainedHeaderBlock chainedHeaderBlock in this.pendingBlocksCache.Values)
-                {
-                    Transaction tx = chainedHeaderBlock.Block.Transactions.FirstOrDefault(x => x.GetHash() == trxid);
+            uint256 blockId = await this.GetBlockIdByTransactionIdAsync(trxid).ConfigureAwait(false);
+            if (blockId == null)
+                return default(Transaction);
 
-                    if (tx != null)
-                    {
-                        this.logger.LogTrace("Transaction '{0}' was found in the pending blocks cache.", trxid);
-                        return Task.FromResult(tx);
-                    }
-                }
-            }
+            Block block = await this.GetBlockAsync(blockId).ConfigureAwait(false);
 
-            return this.blockRepository.GetTransactionByIdAsync(trxid);
+            return block?.Transactions.FirstOrDefault(x => x.GetHash() == trxid);
         }
 
         /// <inheritdoc/>
-        public Task<uint256> GetBlockIdByTransactionIdAsync(uint256 trxid)
+        public async Task<uint256> GetBlockIdByTransactionIdAsync(uint256 trxid)
         {
             lock (this.blocksCacheLock)
             {
-                foreach (ChainedHeaderBlock chainedHeaderBlock in this.pendingBlocksCache.Values)
+                // According to SaveBlockAsync the last block in the list is considered to be on the current main chain.
+                if (this.chain.GetBlock(this.BlockStoreCacheTip.HashBlock) != null)
```

Comment:
I disagree that we need to check the store tip is on mainchain, it  solves a very small race condition here that is anyway still a problem if there is a reorg

Reply:
The goal of the change was to enforce the principle that a caching mechanism should be transparent and not affect the results.

Currently timing considerations as to when the asynchronous `DequeueBlocksContinuouslyAsync` executes, changes where blocks are found (cache vs repo), which leads to different results:  Data in the block repository is non-ambiguous, and on the main chain, whilst the cache is accessed in an ambiguous fashion.